engine/common/rpc/execution_node_identities_provider.go

```diff
@@ -131,7 +137,52 @@ func (e *ExecutionNodeIdentitiesProvider) ExecutionNodesForBlockID(
 	}
 
 	if len(subsetENs) == 0 {
-		return nil, fmt.Errorf("no matching execution node found for block ID %v", blockID)
+		return nil, ErrNoENsFoundForExecutionResult
+	}
+
+	return subsetENs, nil
+}
+
+// ExecutionNodesForResultID returns execution node identities that produced receipts
+// for the specific execution result ID within the given block.
+//
+// Expected errors during normal operation:
+//   - storage.ErrNotFound - if no execution receipts were found in storage for the given block
+//   - ErrNoENsFoundForExecutionResult - if no execution nodes were found that produced
+//     the provided execution result and matched the operators criteria
+func (e *ExecutionNodeIdentitiesProvider) ExecutionNodesForResultID(
+	blockID flow.Identifier,
+	resultID flow.Identifier,
+) (flow.IdentitySkeletonList, error) {
+	allReceipts, err := e.executionReceipts.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve execution receipts for block ID %v: %w", blockID, err)
+	}
+
+	executionReceiptMetaList := make(flow.ExecutionReceiptMetaList, 0, len(allReceipts))
+	for _, r := range allReceipts {
+		executionReceiptMetaList = append(executionReceiptMetaList, r.Meta())
+	}
+
+	receiptsByResultID := executionReceiptMetaList.GroupByResultID()
+	targetReceipts := receiptsByResultID.GetGroup(resultID)
+
+	if len(targetReceipts) == 0 {
+		return nil, fmt.Errorf("no execution receipts found for result ID %v in block %v", resultID, blockID)
```

Comment:
This is an exception in all cases except when the root result is requested. See the existing handler:
https://github.com/onflow/flow-go/blob/8c30c496cf7b6e3fadad97328697a42f633a4853/engine/common/rpc/execution_node_identities_provider.go#L77-L86

In this case, we could similarly check if the result's block is the root block, and return the full set of executors.